clangd: Implement switch source/header extension

thorbenk opened the pull request:
Release Notes:

- Added switch source/header action for clangd language server (fixes [#12801](https://github.com/zed-industries/zed/issues/12801)).

Note: I'm new to both rust and this codebase. I started my implementation by copying how rust analyzer's "expand macro" LSP extension is implemented. I don't yet understand some of the code I copied (mostly the way to get the `server_to_query` in `clangd_ext.rs` and the whole proto implementation).

thorbenk committed:
clangd lsp: implement switch source/header extension

thorbenk commented:
I'd appreciate some feedback on the approach and some help regarding the protobuf part :sweat_smile:

SomeoneToIgnore commented on crates/project/src/lsp_ext_command.rs:
```diff
@@ -135,3 +135,98 @@ impl LspCommand for ExpandMacro {
+        _: &Arc<LanguageServer>,
+        _: &AppContext,
+    ) -> SwitchSourceHeaderParams {
+        log::info!("SwitchSourceHeader::to_lsp(path={:?})", path);
```
Do we need this?

SomeoneToIgnore commented on crates/proto/src/proto.rs:
```diff
@@ -395,6 +395,9 @@ messages!(
     (SynchronizeContextsResponse, Foreground),
+    (LspExtSwitchSourceHeader, Background),
+    (LspExtSwitchSourceHeaderResponse, Background),
+    // Remote development
```
```suggestion
```

Let's remove extra comments, as most of this list is oddly structured and there were no comments before?

thorbenk replied:
These comments were actually recently removed in 5f7881fc1e and I messed up the rebase :sweat_smile:

SomeoneToIgnore commented on crates/editor/src/clangd_ext.rs:
```diff
@@ -0,0 +1,119 @@
+    _: &SwitchSourceHeader,
+    cx: &mut ViewContext<'_, Editor>,
+) {
+    log::info!("switch_source_header action");
```
This logging looks odd, let's remove it?

SomeoneToIgnore commented on crates/editor/src/clangd_ext.rs:
```diff
@@ -0,0 +1,119 @@
+                .read(cx)
+                .language_servers_for_buffer(buffer.read(cx), cx)
+                .find_map(|(adapter, server)| {
+                    log::info!("adapter.name.0: {:?}", adapter.name.0);
```
This too, feels like all new `log::` calls are debug-ones.
You can use `dbg!` (or `dbg!(&)`) for similar purposes and our Clippy profile will fail when run on CI.

thorbenk replied:
Thanks for pointing out the `dbg!` rules. Yes, I usually leave way too many debug prints in my code :sweat_smile:

SomeoneToIgnore commented on crates/editor/src/clangd_ext.rs:
```diff
@@ -0,0 +1,119 @@
+        .unwrap()
+        .to_owned();
+
+    log::info!("switch source/header: source_file: {:?}", source_file);
```
Here.

SomeoneToIgnore commented on crates/editor/src/clangd_ext.rs:
```diff
@@ -0,0 +1,119 @@
+            .await
+            .context("switch source/header")?;
+        if switch_source_header.0.is_empty() {
+            log::info!("clangd returned an empty string for switch source/header");
```
Here.

thorbenk replied:
This is actually indicates a failure on clangd's side: https://clangd.llvm.org/extensions#switch-between-sourceheader and means that the action the user called will not have any effect.

> If the corresponding file can’t be determined, "" is returned. [bug?](https://github.com/clangd/clangd/issues/12)

So, I guess here at least an info message is warranted? Should I pass the original filepath for which the request was issued through to the result, so it can be printed here?

SomeoneToIgnore replied:
But then it's an error, not the `info` log?
Depending on how popular it is to return an empty string for `clangd`, I'd either do `anyhow::ensure!` here (if rare), or at least use a `warn!` log level (if frequent).

SomeoneToIgnore commented on crates/editor/src/clangd_ext.rs:
```diff
@@ -0,0 +1,119 @@
+    cx.spawn(|_editor, mut cx| async move {
+        let switch_source_header = switch_source_header_task
+            .await
+            .context("switch source/header")?;
```
Instead of all those logs around, let's use `.with_context(|| format!(...))?` form and print the buffer path (`source_file`) and whatever else useful info here, actually: this is a user-visible error that will bubble up straight into logs.

SomeoneToIgnore commented on crates/editor/src/clangd_ext.rs:
```diff
@@ -0,0 +1,119 @@
+            return Ok(());
+        }
+
+        let goto = Url::parse(&switch_source_header.0).context("switch source/header")?;
```
> "switch source/header"

Same `format!` form with a path name and better wording which operation (url parsing) did fail here would be very good.

SomeoneToIgnore commented on crates/editor/src/clangd_ext.rs:
```diff
@@ -0,0 +1,119 @@
+            .update(&mut cx, |workspace, view_cx| {
+                workspace.open_abs_path(PathBuf::from(goto.path()), false, view_cx)
+            })
+            .context("switch source/header")?
```
Here too.

SomeoneToIgnore commented on crates/editor/src/clangd_ext.rs:
```diff
@@ -0,0 +1,119 @@
+            .all_buffers()
+            .iter()
+            .any(|b| match b.read(cx).language() {
+                Some(l) => is_c_language(l),
```
I wonder if we could do a deeper check here (the one in rust_analyzer_ext.rs is sort of sloppy/wrong and relied on a fact that there's a single Rust langserver, whereas for C or C++ that could be much more different).

We can call `language_server_ids_for_buffer` here similar to what is done below (maybe, we should extract that whole thing into a separate function to call twice?) and ensure that the ext action is registered from proper LSP only.

Feel free to do the same for rust-analyzer if you want to.

SomeoneToIgnore reviewed (commented):
Thank you for the PR.

Please don't get overwhelmed by a number of commentaries, they are mostly about a single issue with the logging approach.
Currently, logs are printed on every command run with the information which Zed technically could log on LSP level if needed.
On the other hand, errors were provided with a very little context which is not good.
I propose to write something detailed in every `.context` (turning them all into `.with_context`) and removing all new `log::` additions.

Besides that, let's ensure we register the action only for `clangd` (and, ideally, only `rust-analyzer` too) and we're ready to merge this PR.

-----------------------

On a general note, I thin what we're doing now should belong to https://github.com/zed-industries/extensions

We're not there yet in terms of API, so some vital things like these are welcome, but let's not bloat those ext requests too much.

thorbenk committed:
clangd lsp: address logging/error-related review comments

thorbenk commented:
@SomeoneToIgnore : Thanks for your detailed review :smiley: I did not know about `with_context`, thanks!

I just did a first pass to clean up most of the logging (addressing most of your comments).

I will do a second pass later (hopefully by the weekend) and try to refactor the command registration conditional on the LSP.

Regarding the protobuf part, I assume this is for remoting? How can I test that part?

SomeoneToIgnore commented:
> Regarding the protobuf part, I assume this is for remoting? How can I test that part?

I think that will work automagically, but here's a somewhat outdated instruction set if you are interested to test it manually: https://github.com/zed-industries/zed/blob/main/docs/src/development/local-collaboration.md

The idea is to have a `postgres` and `foreman` installed, then start postgres, `foreman start` and ` ./script/zed-local -2` will open 2 Zeds inside a local collaboration set-up.

thorbenk committed:
clangd lsp: reuse language server check across clangd and rust-analyzer ext

thorbenk commented:
Hi @SomeoneToIgnore , sorry for the delay. I rebased and pushed the remaining refactoring you were suggesting. Is this what you had in mind?

Regarding testing the colloboration part, I ran `script/bootstrap` on linux, but it fails when running `sqlx database create` with

```
error: error returned from database: password authentication failed for user "postgres"
```

I don't know anything about postgres, and couldn't find an easy way to fix this...

SomeoneToIgnore committed:
Merge branch 'main' into clangd-switch-source-header

SomeoneToIgnore reviewed (approved):
Sorry for not coming earlier, we've had a quite concentration-demanding sprint towards recent large release.

Great generalization, I think the code does logically what it should now, thank you for the fixes.

> error: error returned from database: password authentication failed for user "postgres"

Looks like the postgres was up and accepted the initial handshake, but did not allow to use postgres/postgres login pair due to being configured differently.
This is a good sign, meaning that whatever runs on that port locally in your system is relatively secure.

For tests, something more ephemeral with less string config is usually used, e.g. `Postgres.app` on mac, or `docker compose up postgres` in the root of the repo for people with `Docker`.
I do the latter and it works for me, but in your case you might want to change the port or temporarily stop the already running instance.

I have checked the remote part, and looks like there's a more fundamental issue I have to fix later: `language_servers_for_buffer` returns nothing for client, remote projects.
Turns out Rust's macro expansion command was not doing anything in remote projects before, but was registered and visible in the command palette — your new code made the registration correct, and now there are no actions and nothing related is shown in the command palette.
This looks like a good improvement at this stage 🙂 

I will consider ways to make it better, but not right now.
Meanwhile, this PR works and streamlines the registration code, thank you for fixing all things.
